design/authenticators/resource-restrictions/resource-restrictions.md

```diff
@@ -0,0 +1,350 @@
+    - Here we validate that the `authn-azure` annotations are known. For
+      example, the annotation `authn-azure/subscription-id` is valid, whereas
+      the annotation `authn-azure/unknown` is invalid. We do that to minimize
+      errors from the user, where they will define, for example, the
+      annotation `authn-azure/subscription-ir` (note the typo). In this
+      case we'd rather fail the authentication and point them to their
+      mistake rather than ignore it, and authenticate the request in a weaker
+       granularity than what is expected by the user.
+  - Validate required restrictions exist
+    - For example, in the Azure Authenticator we require hosts to have the
+     `subscription-id` and the `resource-group` annotations. In the K8s
+      Authenticator we require the `namespace` annotation.
+  - Validate mutually exclusive restrictions
+    - For example, in the Azure Authenticator we allow hosts to have
+     only one of the `user-assigned-identity` and the `system-assigned-identity`
+     annotations. In the K8s we allow hosts to have only one of
+      the annotations `deployment`, `deployment-config` & `stateful-set`.
+
+Also here we can see some generalization.
+
+These need to be defined for each authenticator.
+To simplify that, we can define a simple class for each of these constraints.
+The whole constraints will be an array of these constraints, defined once for each authenticator.
+Then, to validate the Resource Restrictions, it only needs to loop on these constraints, and invoke
+`constraint.validate(resource_restrictions)` on each.
+
+This way, adding or changing the logic is more simple and intuitive, while maintaining readability.
+
+Each of the constraints classes will have the same `validate` signature, and will look similar to:
+
+```ruby
+class RequiredConstraint
+
+  def initialize(required:)
+    @required = required
+  end
+
+  def validate(resource_restrictions:)
+    # implement the constraint
+  end
+
+end
+```
+
+It will be used by each authenticator to define how to validate the configured
+restrictions. These are the ones that were defined in the policy for the host,
+regardless of the request.
+
+Example how it will look like for `authn-azure`:
+
+```ruby
+REQUIRED = [SUBSCRIPTION_ID, RESOURCE_GROUP]
+IDENTITY_EXCLUSIVE = [USER_ASSIGNED_IDENTITY, SYSTEM_ASSIGNED_IDENTITY]
+PERMITTED = REQUIRED + IDENTITY_EXCLUSIVE
+
+CONSTRAINTS =
+    [
+        RequiredConstraint.new(required: REQUIRED),
+        PermittedConstraint.new(permitted: PERMITTED),
+        ExclusiveConstraint.new(exclusive: IDENTITY_EXCLUSIVE)
+    ]
+```
+
+The general class will be defined as follows:
+
+- input
+  - An array of constraints, command classes that accept `resource_restrictions` as input.
+  - The `resource_restrictions` hash built in the previous step
+- output
+  - Nothing, relevant error raised if it failed
+
+So the class signature is:
+
+```ruby
+ValidateResourceRestrictions = CommandClass.new(
```

Comment:
i think that `ValidateResourceRestrictionsConfiguration` is a better naming.

Reply:
removed